### DISCUSSION TITLE
Timeout when processing pdf in Github Action

### DISCUSSION THREAD

**Discussion by codingluke** [marp-team/marp#463] [Q&A]
Dear Marp Community
I use marp-cli since 2 years together with github actions. Since lately I experiance some timeout errors when generating pdfs. Locally it works perfect and fast. Does anybody expericance the same?

- Link to the [deployment.yaml](https://github.com/codingluke/bbzbl-modul-319/blob/main/.github/workflows/deploy.yml)
- Linkt to the [package.json](https://github.com/codingluke/bbzbl-modul-319/blob/main/package.json)

THX a lot!

![Screenshot_2023-08-02_000061](https://github.com/marp-team/marp/assets/643436/84f83ea7-8197-4c05-9ff6-761313e7e490)

**Comment by codingluke** [comment]
@yhatt I could actually reproduce this behaviour with the official marp-cli docker image locally.

With the old PUPPETEER_HEADLESS_MODE it fails caused by headless false or so...

```bash
docker run --rm --init -v $PWD:/home/marp/app/ -e LANG=$LANG -e marpteam/marp-cli operatoren.md --allow-local-files --pdf
[  INFO ] Converting 1 markdown...
[  WARN ] Insecure local file accessing is enabled for conversion from
          operatoren.md.
[ ERROR ] Failed converting Markdown. (Protocol error (Page.printToPDF):
          Printing failed)
```

With the NEW PUPPETEER_HEADLESS_MODE it fails caused by the timeout, like in the github actions.

```bash
docker run --rm --init -v $PWD:/home/marp/app/ -e LANG=$LANG -e PUPPETEER_HEADLESS_MODE=new marpteam/marp-cli operatoren.md --allow-local-files --pdf
[  INFO ] Converting 1 markdown...
[  WARN ] Insecure local file accessing is enabled for conversion from
          operatoren.md.
[ ERROR ] Failed converting Markdown. (waiting for Page.printToPDF failed:
          timeout 30000ms exceeded)
```

locally outside a docker image it works perfectly.

**Reply by yhatt** [reply]
The protocol error should have been fixed in the `latest` image. It seems to be brought by (incompatible) major update of [harfbuzz](https://github.com/harfbuzz/harfbuzz): https://github.com/marp-team/marp-cli/pull/539

**Comment by codingluke** [comment]
I just figured out some interesting use-cases:

1. I use the [kroki plugin](https://gist.github.com/yhatt/8931dd98769bcc1c5b4b92fd1234a190), which makes requests to kroki.io, without the mermaid graphs it works for the operatoren.md deck
2. In another deck I reference an image by http and not locally. This deck also has a timeout. When removing the external image it works.

It seems to me that accessing external ressources are problematic with puppeteer. The html file generation works fine.

This all used to work for 2 years.

**Comment by codingluke** [comment]
Ok, here I can prove, **github actions somehow are blocking external requests**, a simple curl to the image I referene is also not working.

```bash
Run curl https://lerneprogrammieren.de/wordpress/wp-content/uploads/uebersicht-programmiersprachen.png --output test.png
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:--  0:00:01 --:--:--     0
  0     0    0     0    0     0      0      0 --:--:--  0:00:02 --:--:--     0
  0     0    0     0    0     0      0      0 --:--:--  0:00:03 --:--:--     0
  0     0    0     0    0     0      0      0 --:--:--  0:00:04 --:--:--     0
  
 ...
 
  0     0    0     0    0     0      0      0 --:--:--  0:02:09 --:--:--     0
  0     0    0     0    0     0      0      0 --:--:--  0:02:10 --:--:--     0
  0     0    0     0    0     0      0      0 --:--:--  0:02:10 --:--:--     0
curl: (28) Failed to connect to lerneprogrammieren.de port 443 after 130573 ms: Connection timed out
Error: Process completed with exit code 28.

```

**Comment by codingluke** [comment]
It seems to be github actions running in [different regions](https://github.com/actions/runner-images/issues/4700#issuecomment-1008718157). After re-running the curl worked...

![Screenshot_2023-08-04_000064](https://github.com/marp-team/marp/assets/643436/b8fede17-c615-43e5-849d-c1fecd96d1af)

Very strange, I close this discussion and will have an 👀 on this.
